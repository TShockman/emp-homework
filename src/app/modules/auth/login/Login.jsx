import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Button, Form, FormGroup, Label, Input, Card} from 'reactstrap';

export default class Login extends Component{
  static propTypes = {
    requestLogin: PropTypes.func.isRequired,
    id: PropTypes.number
  };

  constructor() {
    super();
    this.state = {
      username: '',
      password: ''
    };
  }

  handleFormUpdate = event => {
    event.stopPropagation();
    const {target} = event;
    this.setState({[target.id]: target.value});
  };

  handleSubmit = event => {
    event.stopPropagation();
    const {requestLogin} = this.props;
    const {username, password} = this.state;

    requestLogin({
      username,
      password
    });
  };

  render() {
    const {id} = this.props;
    if (id !== null) {
      // TODO
      console.log("Will need to redirect to the proper location here.")
    }

    return (
        <Card>
          <Form>
            <FormGroup>
              <Label for="username">Username</Label>
              <Input onChange={this.handleFormUpdate} id="username" name="username" placeholder="Jane_Doe"/>
            </FormGroup>
            <FormGroup>
              <Label for="password">Password</Label>
              <Input onChange={this.handleFormUpdate} id="password" name="password" type="password" placeholder="***********"/>
            </FormGroup>
            <Button className="btn btn-primary" onClick={this.handleSubmit}>Login</Button>
          </Form>
        </Card>
    );
  }
}