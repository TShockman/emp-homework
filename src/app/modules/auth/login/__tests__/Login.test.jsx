import React from 'react';
import { mount } from 'enzyme';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import Login from '../Login';

const renderer = new ReactShallowRenderer();

describe('Login', () => {
  it('Renders consistently', () => {
    const props = {
      requestLogin: jest.fn(),
      id: null
    };
    expect(renderer.render(<Login {...props}/>)).toMatchSnapshot();
  });
  it('Updates volatile state on form update', () => {
    const props = {
      requestLogin: jest.fn(),
      id: null
    };
    const wrapper = mount(<Login {...props}/>);
    expect(wrapper.state().username).toBe('');

    const mockEvent = {
      stopPropagation: jest.fn(),
      target: {
        id: 'username',
        value: 'mockuser'
      }
    };
    wrapper.instance().handleFormUpdate(mockEvent);
    expect(mockEvent.stopPropagation).toHaveBeenCalledTimes(1);
    expect(wrapper.state().username).toBe('mockuser');
    wrapper.unmount();
  });
  it('Requests login on form submission', () => {
    const props = {
      requestLogin: jest.fn(),
      id: null
    };
    const wrapper = mount(<Login {...props}/>);
    wrapper.instance().setState({username: 'jondo', password: '12345'});

    const mockEvent = {
      stopPropagation: jest.fn()
    };
    wrapper.instance().handleSubmit(mockEvent);
    expect(mockEvent.stopPropagation).toHaveBeenCalledTimes(1);
    expect(props.requestLogin).toHaveBeenCalledWith({username: 'jondo', password: '12345'});
  });
});