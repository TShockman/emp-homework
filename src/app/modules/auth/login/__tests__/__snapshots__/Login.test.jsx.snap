// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Renders consistently 1`] = `
<div
  className="container-fluid login"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <form
        className="card"
        onSubmit={[Function]}
      >
        <h2>
          Login
        </h2>
        <div
          className="form-group"
        >
          <label
            htmlFor="username"
          >
            Username
          </label>
          <input
            className="form-control"
            id="username"
            name="username"
            onChange={[Function]}
            placeholder="Jane_Doe"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="password"
          >
            Password
          </label>
          <input
            className="form-control"
            id="password"
            name="password"
            onChange={[Function]}
            placeholder="***********"
            type="password"
          />
        </div>
        <button
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="card"
      >
        <span>
          Don't have an account? 
          <a
            href="#"
          >
            Sign Up
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;
